src/Raven.Server/Documents/Handlers/ShardedDocumentHandler.cs

```diff
@@ -25,29 +35,128 @@ public async Task Get()
 
-            using (ContextPool.AllocateOperationContext(out TransactionOperationContext context))
+            try
             {
+                using (ContextPool.AllocateOperationContext(out TransactionOperationContext context))
+                {
 
-                var id = ids.Single(); // TODO: make it work with multiple shards
+                    var shards = ShardLocator.GetDocumentIdsShards(ids, ShardedContext, context);
+                    foreach (var shard in shards.ShardsToIds)
+                    {
 
-                var shardId = ShardedContext.GetShardId(context, id);
+                        UrlStringBuilder.Clear();
+                        UrlStringBuilder.Append("/docs?");
+                        foreach (var id in shard.Value)
+                        {
+                            UrlStringBuilder.Append($"&id={Uri.EscapeUriString(id)}");
+                        }
 
-                var index = ShardedContext.GetShardIndex(shardId);
+                        var cmd = new ShardedCommand {Method = HttpMethod.Get, Url = UrlStringBuilder.ToString()};
+                        ContextPool.AllocateOperationContext(out TransactionOperationContext ctx);
+                        CmdsContext.Add(ctx);
 
-                var cmd = new ShardedCommand
+                        var task = ShardedContext.RequestExecutors[shard.Key].ExecuteAsync(cmd, ctx);
+                        Task.Add(task);
+                        Cmds.Add(shard.Key, cmd);
+                    }
+
+                    await System.Threading.Tasks.Task.WhenAll(Task);
+
+                    HttpContext.Response.StatusCode = (int)ShardedStatusCodeChooser.GetStatusCode(Cmds.Values);
+
+                    if (HttpContext.Response.IsSuccessStatusCode() == false)
+                        return;
+
+                    var actualEtag = ComputeHttpEtags.CombineEtags(EnumerateEtags(Cmds));
+
+                    if (etag == actualEtag)
+                    {
+                        HttpContext.Response.StatusCode = (int)HttpStatusCode.NotModified;
+                        return;
+                    }
+
+                    HttpContext.Response.Headers[Constants.Headers.Etag] = "\"" + actualEtag + "\"";
+
+                    using (var writer = new AsyncBlittableJsonTextWriter(context, ResponseBodyStream(), ServerStore.ServerShutdown))
+                    {
+                        writer.WriteStartObject();
+
+                        WriteArray(writer, nameof(GetDocumentsResult.Results), ids, shards);
+
+                        //TODO: write includes
+
+                        writer.WriteEndObject();
+                        await writer.OuterFlushAsync();
+                        //TODO: Add performance hints
+                    }
+                }
+            }
+            finally
+            {
+                Task.Clear();
+                Cmds.Clear();
+
+                Parallel.ForEach(CmdsContext, c =>
```

Comment:
That is not a good idea, we want to return this to the same thread as we pull this from